File: tests/integration_tests/adapters/betfair/test_orderbook.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2021 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------

from nautilus_trader.adapters.betfair.data import on_market_update
from nautilus_trader.model.data.tick import TradeTick
from nautilus_trader.model.data.venue import InstrumentClosePrice
from nautilus_trader.model.data.venue import InstrumentStatusUpdate
from nautilus_trader.model.orderbook.book import L2OrderBook
from nautilus_trader.model.orderbook.data import OrderBookDelta
from nautilus_trader.model.orderbook.data import OrderBookDeltas
from nautilus_trader.model.orderbook.data import OrderBookSnapshot
from tests.integration_tests.adapters.betfair.test_kit import BetfairDataProvider
from tests.integration_tests.adapters.betfair.test_kit import BetfairTestStubs


def test_betfair_orderbook(betfair_data_client, provider):
    provider.load_all()
    book = L2OrderBook(
        instrument_id=BetfairTestStubs.instrument_id(),
        price_precision=2,
        size_precision=2,
    )
    for update in BetfairDataProvider.raw_market_updates():
        for message in on_market_update(instrument_provider=provider, update=update):
            try:
                if isinstance(message, OrderBookSnapshot):
                    book.apply_snapshot(message)
                elif isinstance(message, OrderBookDeltas):
                    book.apply_deltas(message)
                elif isinstance(message, OrderBookDelta):
                    book.apply_delta(message)
                elif isinstance(message, (TradeTick, InstrumentStatusUpdate, InstrumentClosePrice)):
                    pass
                else:
                    raise NotImplementedError(str(type(message)))
            except Exception as ex:
                print(str(type(ex)) + " " + str(ex))

File: crates/model/tests/test_order_book.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

use std::path::Path;

use nautilus_core::paths::get_workspace_root_path;
use nautilus_databento::loader::DatabentoDataLoader;
use nautilus_model::{enums::BookType, identifiers::InstrumentId, orderbook::OrderBook};
use nautilus_testkit::{
    common::{get_test_data_file_path, get_test_data_large_checksums_filepath},
    files::ensure_file_exists_or_download_http,
};
use rstest::*;

#[rstest]
pub fn test_order_book_databento_mbo_nasdaq() {
    let checksums = get_test_data_large_checksums_filepath();
    let filename = "databento_mbo_xnas_itch.csv";
    let file_path = get_test_data_file_path(format!("large/{filename}").as_str());
    let url = "https://hist.databento.com/v0/dataset/sample/download/xnas.itch/mbo";
    ensure_file_exists_or_download_http(Path::new(file_path.as_str()), url, Some(&checksums))
        .unwrap();

    let instrument_id = InstrumentId::from("AAPL.XNAS");
    let mut _book = OrderBook::new(instrument_id, BookType::L3_MBO);

    let publishers_filepath = get_workspace_root_path()
        .join("adapters")
        .join("databento")
        .join("publishers.json");
    let _loader = DatabentoDataLoader::new(Some(publishers_filepath)).unwrap();
    // let deltas = loader
    //     .load_order_book_deltas(filepath, Some(instrument_id))
    //     .unwrap();
    //
    // for delta in deltas.iter() {
    //     book.apply_delta(delta);
    // }

    // assert_eq!(book.best_bid_price().unwrap(), price);
    // assert_eq!(book.best_ask_price().unwrap(), price);
    // assert_eq!(book.best_bid_size().unwrap(), size);
    // assert_eq!(book.best_ask_size().unwrap(), size);
}
